README.md
# 🧾✨ Invoice Magic Wand for Laravel

Unleash the alchemy of billing with Invoice Magic Wand for Laravel! 🪄📜

This mystical package conjures up invoices with a flick of your developer's wrist, transforming the mundane task of invoice generation into a delightful display of digital wizardry. Whether you're billing with **Billingo** or summoning statements with **Szamlazz.hu**, this enchanting toolkit makes the numbers dance to your tune. No more grunt work, just sparkling invoices popping out like rabbits from a magician's hat! 🐇💫

## 📦 Installation

To install the package, run the following command in your Laravel project:

```bash
composer require composite/invoice-wrapper
```
Publish the configuration file with:

```bash
php artisan vendor:publish --provider="Composite\InvoiceWrapper\InvoiceWrapperServiceProvider"
```

## ⚙️ Configuration
After publishing the config file, you should set your environment variables in your .env file:
```dotenv
INVOICING_PROVIDER=billingo

# Billingo Configuration
BILLINGO_BASE_URL=https://api.billingo.hu/v3/
BILLINGO_API_KEY=your-billingo-api-key
BILLINGO_BLOCK_ID=your-billingo-block-id

# Szamlazz.hu Configuration
SZAMLAZZHU_API_KEY=your-szamlazzhu-api-key
```
Make sure to replace your-billingo-api-key, your-billingo-block-id, and your-szamlazzhu-api-key with your actual API keys and configuration details.

## 🛠 Usage
To issue an invoice, you can use the InvoiceWrapper facade with the desired invoice payload. Here's an example:

```php
use Composite\InvoiceWrapper\Facades\InvoiceWrapper;
use Carbon\Carbon;

$invoicePayload = [
    // Your invoice payload here
];

// Issue the invoice using the selected provider
$invoice = InvoiceWrapper::issueInvoice($invoicePayload);
```

or without Facade, as a service (it can be useful, when you want to use multiple providers):

```php
use Composite\InvoiceWrapper\InvoiceWrapper;
use Carbon\Carbon;

$config = [
    'selected_provider' => env('INVOICING_PROVIDER'),
    'providers' => [
        'billingo' => [
            'name' => 'Billingo',
            'base_url' => env('BILLINGO_BASE_URL', 'https://api.billingo.hu/v3/'),
            'api_key' => env('BILLINGO_API_KEY'),
            'block_id' => env('BILLINGO_BLOCK_ID', 0),
        ],
        'szamlazzhu' => [
            'name' => 'Szamlazz.hu',
            'api_key' => env('SZAMLAZZHU_API_KEY'),
        ]
    ],
];

$invoicePayload = [
    // Your invoice payload here
];

// Issue the invoice using the selected provider
$invoiceWrapper = new InvoiceWrapper(
 \Composite\InvoiceWrapper\Factories\InvoiceGatewayFactory::create($config)
);

$invoiceWrapper->issueInvoice($invoicePayload);
```

Sample invoice payload:

```php
$invoicePayload = [
    'partner' => [
        'id' => null, // for billingo required
        'name' => 'Test Partner',
        'address' => [
            'country_code' => 'HU',
            'post_code' => '1111',
            'city' => 'Budapest',
            'address' => 'Test utca 1.',
        ],
        'tax_type' => 'NO_TAX_NUMBER', //"HAS_TAX_NUMBER" : "NO_TAX_NUMBER",
        'taxcode' => 'HU29168950',
        'email' => 'btamba@composite.hu', // if email sending is true it is required
        'send_email' => true,
    ],
    'invoice' => [
        // 'type' => 'invoice', // not handled yet
        'fulfillment_date' => Carbon::now()->format('Y-m-d'),
        'due_date' => Carbon::now()->format('Y-m-d'),
        'payment_method' => 'bankcard', //'bankcard','cash','wire_transfer','cash_on_delivery','paypal','szep_card',
        'language' => 'hu', // 'hu', 'en'
        'currency' => 'HUF', // 'HUF', 'EUR'
        'paid' => false,
        'items' => [
            [
                'name' => 'Test product',
                'unit_price' => 1000,
                'unit_price_type' => 'net', // net, gross
                'quantity' => 1,
                'unit' => 'db',
                'vat' => '5', // '0%','5%','18%','27%','27%'
            ],
            [
                'name' => 'Test product 2',
                'unit_price' => 2000,
                'unit_price_type' => 'gross', // net, gross
                'quantity' => 1,
                'unit' => 'db',
                'vat' => '27', // '0%','5%','18%','27%','27%'
            ],
            [
                'name' => 'Test product 3',
                'unit_price' => 2000,
                'unit_price_type' => 'net', // net, gross
                'quantity' => 2,
                'unit' => 'db',
                'vat' => '27', // '0%','5%','18%','27%','27%'
            ]
        ],
        // 'conversion_rate' => 1, // Not handled yet
        'comment' => 'It is a comment',
    ],
];
```

To retrieve an invoice:

```php
$invoiceId = 'invoice-id-here';
$invoice = InvoiceWrapper::getInvoice($invoiceId);
```

To download an invoice:

```php
$invoiceId = 'invoice-id-here';
$invoice = InvoiceWrapper::downloadInvoice($invoiceId);
```

Make sure to handle any exceptions that may be thrown due to API errors or configuration issues.

## 📝✨To-Do List for Future Enhancements
Here's a whimsical checklist of tasks for the Invoice Wrapper's journey ahead:

- [x] 🧙‍️ Creating the Invoice Wrapper - Done!
- [x] 🧞‍ Implementing Szamlazz.hu Service - Done!
- [x] 🧞‍ Implementing Billingo Service - "The Genie is out of the bottle!"
- [x] 📩 Downloading Invoice - Awaiting the magic spell!
- [ ] 🔗 Getting Download Links - Soon to be summoned from the digital ether!

## 🤝 Support
For issues, questions, and contributions, please use the GitHub issues section of this repository.

## 🌟 Credits
This package is proudly brought to you by:

- 🏢 [Composite Solutions](https://github.com/Composite-Solutions) - For innovative solutions and dedicated support.
- 👨‍💻 [Bence Tamba](https://github.com/tambabence) - The mastermind behind the code magic.

A huge shout-out to all contributors and supporters! Your feedback and contributions make this project better every day. 🚀🙌

Thank you for being part of this journey! 🎉👏

## 📜 Changelog
All notable changes to this project will be documented in the [CHANGELOG.md](CHANGELOG.md) file.

## 📜 License
This package is open-source software licensed under the [MIT license](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.2] - 2023-11-13

### Fixed
- Szamlazz.hu trait payload fix

## [1.1.1] - 2023-11-13

### Fixed
- Illuminate dependency version fix

## [1.1.0] - 2023-11-13

### Added
- Szamlazz.hu invoice download feature
- Billingo service

## [1.0.0] - 2023-11-10

### Added
- Initial release
- Szamlazz.hu service
- Invoice creation
- Invoice retrieval


src/Traits/SzamazzhuHelper.php
<?php

namespace Composite\InvoiceWrapper\Traits;

use SzamlaAgent\Currency;
use SzamlaAgent\Document\Document;
use SzamlaAgent\Language;
use SzamlaAgent\TaxPayer;

trait SzamazzhuHelper
{
    /**
     * @param array $invoiceResponse
     * @return array
     */
    private function formatInvoiceResponse(array $invoiceResponse): array
    {
        $items = array_map(function ($item) {
            return $this->formatItem($item);
        }, $invoiceResponse['tetelek']);

        return [
            'partner' => $this->formatPartner($invoiceResponse['vevo']),
            'invoice' => [
                'invoice_id' => $invoiceResponse['alap']['szamlaszam'],
                'invoice_number' => $invoiceResponse['alap']['szamlaszam'],
                'fulfillment_date' => $invoiceResponse['alap']['telj'],
                'due_date' => $invoiceResponse['alap']['fizh'],
                'payment_method' => $this->getPaymentMethodType($invoiceResponse['alap']['fizmod']),
                'language' => $invoiceResponse['alap']['nyelv'],
                'currency' => $invoiceResponse['alap']['devizanem'],
                // 'paid' => $this->getPaymentStatus($invoiceResponse), // Assuming a method to determine this
                'items' => $items,
                'comment' => $invoiceResponse['alap']['megjegyzes'],
            ],
        ];
    }

    /**
     * @param array $partner
     * @return array
     */
    private function formatPartner(array $partner): array
    {
        return [
            'id' => (int)$partner['id'],
            'name' => $partner['nev'],
            'address' => $this->formatAddress($partner['cim']),
            'taxcode' => $partner['adoszam'],
            'email' => $partner['email'],
        ];
    }

    /**
     * @param array $address
     * @return array
     */
    private function formatAddress(array $address): array
    {
        return [
            'country_code' => $this->getCountryCode($address['orszag'] ?? ''),
            'post_code' => $address['irsz'],
            'city' => $address['telepules'],
            'address' => $address['cim'],
        ];
    }

    /**
     * @param array $item
     * @return array
     */
    private function formatItem(array $item): array
    {
        return [
            'name' => $item['nev'],
            'unit_net_price' => $item['nettoegysegar'],
            'quantity' => $item['mennyiseg'],
            'unit' => $item['mennyisegiegyseg'],
            'vat' => $item['afakulcs'],
            'net_price' => $item['netto'],
            'vat_price' => $item['afa'],
            'gross_price' => $item['brutto'],
        ];
    }

    /**
     * @param $method
     * @return string
     */
    private function getPaymentMethod($method): string
    {
        return match ($method) {
            'bankcard' => Document::PAYMENT_METHOD_BANKCARD,
            'cash' => Document::PAYMENT_METHOD_CASH,
            'wire_transfer' => Document::PAYMENT_METHOD_TRANSFER,
            'cash_on_delivery' => Document::PAYMENT_METHOD_CASH_ON_DELIVERY,
            'paypal' => Document::PAYMENT_METHOD_PAYPAL,
            'szep_card' => Document::PAYMENT_METHOD_SZEP_CARD,
            default => Document::PAYMENT_METHOD_CASH,
        };
    }

    /**
     * @param $method
     * @return string
     */
    private function getPaymentMethodType($method): string
    {
        return match ($method) {
            Document::PAYMENT_METHOD_BANKCARD => 'bankcard',
            Document::PAYMENT_METHOD_CASH => 'cash',
            Document::PAYMENT_METHOD_TRANSFER => 'wire_transfer',
            Document::PAYMENT_METHOD_CASH_ON_DELIVERY => 'cash_on_delivery',
            Document::PAYMENT_METHOD_PAYPAL => 'paypal',
            Document::PAYMENT_METHOD_SZEP_CARD => 'szep_card',
            default => 'cash',
        };
    }

    /**
     * @param $currency
     * @return string
     */
    private function getCurrency($currency): string
    {
        return match ($currency) {
            'HUF' => Currency::CURRENCY_HUF,
            'EUR' => Currency::CURRENCY_EUR,
            default => Currency::CURRENCY_HUF,
        };
    }

    /**
     * @param $language
     * @return string|null
     */
    private function getLanguage($language): ?string
    {
        return match ($language) {
            'hu' => Language::LANGUAGE_HU,
            'en' => Language::LANGUAGE_EN,
            default => null,
        };
    }

    /**
     * @param $taxType
     * @return int
     */
    private function getTaxPayer($taxType): int
    {
        return match ($taxType) {
            'HAS_TAX_NUMBER' => TaxPayer::TAXPAYER_HAS_TAXNUMBER,
            default => TaxPayer::TAXPAYER_NO_TAXNUMBER,
        };
    }

    /**
     * @param $vatRate
     * @return string
     */
    private function getVat($vatRate): string
    {
        return match ($vatRate) {
            '5', '18', '27', '0' => $vatRate,
            default => '27',
        };
    }

    /**
     * @param $countryName
     * @return string
     */
    private function getCountryCode($countryName): string
    {
        return match ($countryName) {
            'Magyarország' => 'HU', // TODO: get from config, add more countries
            default => $countryName,
        };
    }

    /**
     * @param $countryCode
     * @return string
     */
    private function getCountryName($countryCode): string
    {
        return match ($countryCode) {
            'HU' => 'Magyarország', // TODO: get from config, add more countries
            default => $countryCode,
        };
    }
}
